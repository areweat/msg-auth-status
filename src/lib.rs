#![warn(
    clippy::unwrap_used,
//    missing_docs,
    rust_2018_idioms,
    unused_lifetimes,
    unused_qualifications
)]
#![doc = include_str!("../README.md")]

mod parser;
pub use parser::auth_results::{
    AuthenticationResults, DkimResultCode, IpRevResultCode, SpfResultCode,
};

#[derive(Debug)]
pub struct MessageAuthStatus {}

#[derive(Debug)]
pub enum Error {
    ParseNone,
}

use std::borrow::Cow;

impl<'hdr> MessageAuthStatus {
    #[cfg(feature = "mail_parse")]
    pub fn from_rfc822(msg: &'hdr [u8]) -> Result<Self, Error> {
        let parsed = match mail_parser::MessageParser::default().parse(&*msg) {
            Some(p) => p,
            None => return Err(Error::ParseNone),
        };
        let auth_results: Vec<AuthenticationResults<'_>> = parsed
            .header_values("Authentication-Results")
            .map(|mh| mh.into())
            .collect();
        panic!("{:?}", auth_results);
        //panic!("{:?}", parsed.header_values("DKIM-Signature").join(","));

        Ok(Self {})
    }
}

#[cfg(test)]
mod test {
    #[allow(unused_imports)]
    use super::*;
    #[allow(unused_imports)]
    use insta::assert_debug_snapshot;
    #[allow(unused_imports)]
    use rstest::rstest;
    #[allow(unused_imports)]
    use std::{fs::File, io::Read, path::PathBuf};

    #[allow(dead_code)]
    fn load_test_data(file_location: &str) -> Vec<u8> {
        let mut file = File::open(file_location).unwrap();
        let mut data: Vec<u8> = vec![];
        file.read_to_end(&mut data).unwrap();
        data
    }

    #[rstest]
    #[cfg(feature = "mail_parse")]
    fn from_rfc822(#[files("test_data/rfc8601_b*.txt")] file_path: PathBuf) {
        let data = load_test_data(file_path.to_str().unwrap());
        assert_debug_snapshot!(MessageAuthStatus::from_rfc822(&*data));
    }
}
