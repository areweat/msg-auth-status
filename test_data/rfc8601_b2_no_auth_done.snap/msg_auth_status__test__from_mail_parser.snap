---
source: src/lib.rs
expression: "&status"
---
Ok(
    MessageAuthStatus {
        results: [
            AuthenticationResults {
                host: Some(
                    HostVersion {
                        host: "example.org",
                        version: Some(
                            1,
                        ),
                    },
                ),
                smtp_auth_result: [],
                spf_result: [],
                dkim_result: [],
                iprev_result: [],
                none_done: true,
                raw: Some(
                    "example.org 1; none",
                ),
            },
        ],
        dkim_signatures: [],
    },
)
